### PR TITLE
feat: add max-num-batched-tokens configuration and implement request handling constraints (#83)

### PR DESCRIPTION
**Pull request by mohitpalsingh**
## ✨ New Feature: `max-num-batched-tokens` Support

This PR implements the `max-num-batched-tokens` parameter, which limits the total number of tokens (prompt + max output tokens) that can be processed simultaneously across all running requests.

### 🔧 Technical Implementation
* Added `calculateProcessingTokens()` function to compute total token requirements per request.
* Implemented `canAcceptRequest()` function to check constraint satisfaction.
* Added `addRunningRequest()` and `removeRunningRequest()` for proper token tracking.
* Redesigned queue manager with ticker-based processing to avoid busy-waiting.
* Added atomic counters for thread-safe token counting.

### 📝 Configuration & Documentation
* Added command-line parameter: `--max-num-batched-tokens` with proper help text.
* Updated configuration files: Added parameter to `config.yaml` (2048) and `basic-config.yaml` (1024).
* Enhanced README: Added documentation for the new parameter.
* Updated manifests: Configuration examples now include the new parameter.

### 🔄 Code Quality Improvements
* Maintained backward compatibility - when set to `0`, only the `max-num-seqs` constraint is enforced.

---

### 🧪 Testing
* ✅ All existing tests pass (102/102 specs).
* ✅ Added comprehensive test coverage for the new functionality.
* ✅ Test scenarios include:
    * Parameter validation and edge cases
    * Queue constraint enforcement
    * Token calculation accuracy
    * Configuration loading from files
    * Backward compatibility (disabled mode)

---

### 📊 Behavior

**When `max-num-batched-tokens` is configured:**
* Requests are queued until both constraints are satisfied:
    * Number of running requests < `max-num-seqs`
    * Total processing tokens + new request tokens ≤ `max-num-batched-tokens`

**When `max-num-batched-tokens` is `0` or not set:**
* Only the `max-num-seqs` constraint is enforced (existing behavior).

---

### 🏗️ Files Modified
* `config.go` - Added parameter and validation
* `simulator.go` - Core implementation and renamed field
* `pkg/llm-d-inference-sim/*_test.go` - Updated tests and expectations
* `config.yaml` & `basic-config.yaml` - Added parameter
* `README.md` - Documentation updates

---

### ✅ Validation
- [x] Build succeeds without errors
- [x] All tests pass (102/102 specs)
- [x] Parameter appears in `--help` output
- [x] Configuration loading works correctly
- [x] Token tracking is thread-safe
- [x] Backward compatibility maintained

---

### 🎯 Addresses
This implementation follows the `vLLM` specification for the `max-num-batched-tokens` parameter, ensuring requests only proceed when both sequence and token constraints are satisfied. This enables better resource management and throughput control.

### CODE REVIEW

File: pkg/llm-d-inference-sim/config_test.go [hunk 1]
```diff
@@ -88,6 +88,7 @@ var _ = Describe("Simulator configuration", func() {
 	c = createDefaultConfig(qwenModelName)
 	c.Port = 8001
 	c.ServedModelNames = []string{"model1", "model2"}
+	c.MaxNumBatchedTokens = 2048
```
Comment (irar2) [inline pkg/llm-d-inference-sim/config_test.go]: Please move this (and all other occurrences) to createDefaultConfig().

File: pkg/llm-d-inference-sim/simulator.go [hunk 2]
```diff
@@ -375,6 +393,72 @@ func (s *VllmSimulator) isLora(model string) bool {
...
+	// If max_tokens is not specified, calculate it as max-model-len - prompt-len
+	outputTokens := 0
+	if maxCompletionTokens != nil {
+		outputTokens = int(*maxCompletionTokens)
+	} else {
+		outputTokens = s.config.MaxModelLen - promptTokens
+		if outputTokens < 0 {
+			outputTokens = 0
+		}
+	}
```
Comment (irar2) [inline pkg/llm-d-inference-sim/simulator.go:396]: Isn't it that if maxCompletionTokens is nil, this function should just return s.config.MaxModelLen?

File: pkg/llm-d-inference-sim/simulator.go [hunk 3]
```diff
@@ -375,6 +393,72 @@ func (s *VllmSimulator) isLora(model string) bool {
...
+
+// addRunningRequest adds a request to the running requests tracking
+func (s *VllmSimulator) addRunningRequest(reqID string, req completionRequest) {
+	processingTokens := s.calculateProcessingTokens(req)
+
+	runningReq := runningRequest{
+		promptTokens: req.getNumberOfPromptTokens(),
+		maxTokens:    processingTokens,
+		totalTokens:  processingTokens,
+	}
```
Comment (irar2) [inline pkg/llm-d-inference-sim/simulator.go]: I could only find where 'totalTokens' is used (to update processingTokensCount), why 'promptTokens' and 'totalTokens'  are needed? And if they are not used, I guess we don't need runningRequestsMap at all? (And requestID)

**Review by irar2 — CHANGES_REQUESTED** [review]

**Comment by irar2** [comment]
@mohitpalsingh Thank you very much for this PR. After review of this PR along with issue #83 with a colleague of mine, we need to think about what exactly needs to be simulated here more carefully. Therefore, there will be a delay in the review of your PR.

File: pkg/llm-d-inference-sim/config_test.go (hunk 1 above)
Comment (mohitpalsingh) [reply pkg/llm-d-inference-sim/config_test.go]: done

File: pkg/llm-d-inference-sim/simulator.go (hunk 2 above)
Comment (mohitpalsingh) [reply pkg/llm-d-inference-sim/simulator.go:396]: resolved

File: pkg/llm-d-inference-sim/simulator.go (hunk 3 above)
Comment (mohitpalsingh) [reply pkg/llm-d-inference-sim/simulator.go]: You're right, that was unnecessary, I was biased on the general one fit for all use case approach in-case we would like further control over parallel requests in near future. But I guess removing it for now is better and leaner.
 
**Removed unused fields and structures:**
- runningRequest struct - Completely removed since promptTokens and maxTokens were never used
- runningRequestsMap sync.Map - Removed since we don't need to map request IDs to token counts
- requestID field - Removed from completionReqCtx since we no longer need unique request tracking

**Simplified token tracking:**
- Before: Store a complex runningRequest struct with 3 fields in a map, indexed by requestID
- After: Store just the processingTokens count directly in the completionReqCtx

**Updated method signatures:**
- addRunningRequest() now takes *completionReqCtx instead of (reqID, req)
- removeRunningRequest() now takes *completionReqCtx instead of reqID
- Both methods are simpler and more direct
